transformer_engine/pytorch/csrc/thd_utils.cuh

```diff
@@ -183,59 +183,287 @@ __global__ void thd_lse_kernel(lse_dtype *lse, float *half_lse, int *cu_seqlens,
 }
 
 /***************************************************************************************************
- * Support THD format for Context Parallel: Out correction in forward
+ * Support BSHD, SBHD, and THD formats for Context Parallel: Out correction in forward
  **************************************************************************************************/
 
-template <typename dtype, int only_second_half, int tile_size, bool lse_packed>
-__global__ void thd_out_correction_kernel(dtype *out, dtype *out_per_step, float *lse,
-                                          float *lse_per_step, int *cu_seqlens, int batch,
-                                          int num_heads, int dim_per_head, int lse_seqlen,
-                                          int lse_per_step_seqlen) {
+// Stores pointers to output and lse tensors for batch kernel launch.
+template <int n>
+struct TensorList {
+  void *addresses_out[n];
+  void *addresses_lse[n];
+  int num_tensors_this_launch;
+};
```

Comment:
Did you check the performance of this approach? What is the typical number `n` that you are using? A few years ago (not sure if this changed) compiler would dump such struct to local memory as long as it wasn't big enough (128B+). For details please see NVBug 2964767.

Reply:
Hi @ptrendx 

`n` should be equal to CP size, which is typically something like 2, 4, 8, 16, 32, 64, etc, I do not expect it to be very big. What's you suggested change to this?